initial deprecation spec (#205)
kabanero-io/kabanero-foundation: design/javamp-deprecation.md

> # Java Microprofile Application Stack Deprecation

Reviewer: Global comment:  the stack name, in the yaml, isn't "Java Microprofile" but "Eclipse MicroProfileÂ®"   
(with that trademark char maybe not necessary here)

The java-openliberty stack name, is indeed "Open Liberty".

Author: Yeah, the stack-id is java-microprofile.  I'm inclined to leave as-is.

## Key Concepts / Background
- Kabanero hasn't had any guidelines for deprecating an application stack, so this design will describe a best practice for deprecation.

- The Java Microprofile community has refactored the Java cloud native  application stack to provide many improvements -- delivered as a new stack: the Open Liberty application stack.  The Open Liberty application stack has already been featured in Kabanero.  New Java cloud native applications should now leverage the Java Open Liberty application stack.

> - The Java Microprofile application stack maintained in Kabanero is derived from the similar stack at Appsody.  Kabanero will inherit the stack metadata which marks the stack deprecated, as well as the container changes which will print deprecated messages as part of the application start up.  This is a best practice.  The deprecation marker and start up messages indicates to any user of the stack that they should start making plans to change technologies.

Reviewer: Fine..as long as you don't think "inherit" implies something more specific than "copy/merge from the fork" in this context.

Author: Yeah, copy/merge is what was intended.

> - The Java Microprofile application stack will continue to be maintained for period of time as determined by the stack provider.  A minimum of 3 milestones is considered a best practice, although some stack providers may choose a different transitional period.

Reviewer: ```suggestion
- The Java Microprofile application stack will continue to be maintained for a period of time as determined by the stack provider.  A minimum of 3 milestones is considered a best practice, although some stack providers may choose a different transitional period.
```

- The default Kabanero stack-hub index will be updated to remove the Java Microprofile application stack.  Users wishing to continue to leverage the Java Microprofile application stack can add the stack index to their own customized stack-hub for usage with Kabanero.

## User stories

- The Kabanero community would like to have a process for deprecating  application stacks.

- The Java Microprofile application stack provider would like to deprecate the Java Microprofile application stack.

> - As a Java-Microprofile application stack consumer, I would like to have a transition period where I can refactor my application to leverage the Open Liberty applciation stack.

Reviewer: ```suggestion
- As a Java-Microprofile application stack consumer, I would like to have a transition period where I can refactor my application to leverage the Open Liberty application stack.
```

> - As a Java-Microprofile application stack consumer, I would like to have a transition period where I can refactor my application to leverage the Open Liberty applciation stack.

Reviewer: Maybe worth broadening to have a statement for each of Jane/Champ as "stack consumers"?

## As-is

- The Kabanero stack-hub index release includes the Java MicroProfile application stack, which for this milestone will be deprecated.

> - There are two Java cloud native application stacks delivered in Kabanero without clear enough guidance on which to leveage.

Reviewer: Is that saying SpringBoot stacks are not "Java cloud-native" ?   Maybe it doesn't matter since we have the Open Liberty stack as context.

Author: Not intending to say anything at all about Spring Boot stack.

## To-be

- As the Kabanero Java MicroProfile stack provider, I would like to deprecate the existing Java MicroProfile application stack in preference to the Java Open Liberty Application Stack.

- The Kabanero stack-hub index release will exclude the now deprecated Java MicroProfile application stack.

- As a Java MicroProfile application stack consumer, I would like to be able continue to leverage the application stack for a period of time until I can transition my Java application to leverage the Open Liberty Stack.

> - As the Kabanero community, I would like to remove guides exhibiting and documentation which indicates that the Java Microprofile should be used to build Java cloud native applications.

Reviewer: ```suggestion
- As the Kabanero community, I would like to remove guides exhibiting and documentation indicating that the Java Microprofile should be used to build Java cloud native applications.
```

## Main Feature design

### Removal of Java MicroProfile application stack from the default stack hub release

The Kabanero stack-hub release will remove the Java MicroProfile application stack from the index.

### Custom Resource Changes

None.

### Kabanero Operator

None.

## Day 2 Operations

None.

### Kabanero Upgrade Scenarios

#### Kabanero 0.8 to 0.9

There are no configuration upgrade transformations due to this design known at this time.

#### (Kabanero 0.6 to 0.9

There are no configuration upgrade transformations due to this design known at this time.

#### (Kabanero updates from 0.1-0.5)

No support for directly upgrading from these releases.

##  Dependent Projects

### Appsody

> This feature is dependent on deprecation changes made to to the Java MicroProfile stack which Kabanero inherits from the Appsody community. 

Reviewer: Again "inherits"?

Author: Same comment, loosely speaking.
